### ISSUE TITLE
Add validation for a 'readOnly' property

### ISSUE THREAD

**Issue by posener** [go-swagger/go-swagger#936]
## Problem statement

Sending a request with property which is defined as 'readOnly' in the schema should return a 400.
Currently it is accepted.

## Swagger specification

readOnly: boolean:

Relevant only for Schema "properties" definitions. Declares the property as "read only". This means that it MAY be sent as part of a response but MUST NOT be sent as part of the request. Properties marked as readOnly being true SHOULD NOT be in the required list of the defined schema. Default value is false.

## Steps to reproduce

1. Create a POST verb on path / with body with schema that has a readOnly property
2. Send a request that contains this property.
3. You won't get an 400 error

## Environment

swagger version: 0.8.0
go version: go1.7.4
OS: linux/amd64

**Comment by dbaumgarten** [comment]
+1
I also would really like to have this feature. An api I am building has some properties in its models like createdAt or modifiedAt, that are declared as readOnly. Currently I have to "manually" remove these fields from requests, what is quite annoying, and kind of negates the whole purpose of auto-generating code from swagger.

I tried to find the relevant places, where this had to be implemented, but I failed. Although I think I am relatively proficient in go, I have absolutely no idea, what exactly is going on in the go-swagger code.

**Comment by fredbi** [comment]
With go-openapi/validate#50 the spec validate will issue a warning on this situation.
I understand this doesn't fully address the issue, but it's at least a start...

**Comment by fredbi** [comment]
This one is difficult because it makes the validation _contextual_: an instance of a model (or simple schema) may be valid or not depending on who sent this data.

To my knowledge, this is now the **ONLY** swagger 2.0 feature that is not supported by go-swagger.

For simple params and headers, we might find a straightforward way to generate this additional validation, since they only exist in the context of an operation and (at the moment) we never validate responses or response headers. This validation would return an error is the value is not IsZero().

For models representing schemas, the situation is more difficult: the `runtime.Validatable` interface says if a model is valid or invalid, full stop. It is not contextual.

One idea would be to extend the validation capabilities with an additional interface (say, `runtime.ValidatableParameter`, or `ValidatableFor`...) which would carry on the extra check in this specific context (JSON-Schema draft 7 introduces the concept of "owning authority" to describe the newly introduced `readOnly` and `writeOnly`). Or add methods to set a validation context.

I think this one should be discussed to figure out a sound design.

**Comment by gregmarr** [comment]
I think the easiest would be for the validation functionality to have an option to select "read-only data can be included" vs "read-only data must not be included".

```
    type Validatable interface {
        Validate(strfmt.Registry, bool disallowReadOnly) error
    }
```

although if we want to support `writeOnly` as well, then maybe:

    type OperationType int
    const (
        ReadOperation = iota
        WriteOperation
    )
    type Validatable interface {
        Validate(strfmt.Registry, OperationType) error
    }

**Comment by fredbi** [comment]
(good morning Greg).

Mmh yes of course... If you indulge yourself into changing the interface...

The write only is not swagger 2 strictly speaking, but it is inline with json-schema alignment at some (remote) point. anyhow, both specs (openapi and json-schema) does borrow from each other with every new version.

How about something like:

```golang
type ValidatableFor interface {
  Validatable
  WithOperationType(OperationType) Validatable
}
```
So from the outside, we could call validations like: 
```golang
if err := mymodel.WithOperationType(op).Validate(formats) ; err != nil { 
...
```

**Comment by gregmarr** [comment]
I wasn't sure if anyone implemented `Validatable` themselves or if it's always generated.  Since it's in the `runtime` package, I guess we should assume that someone might have implemented it themselves and thus it's an API change that would need to be properly versioned.

How would your wrapper interface work?

**Comment by fredbi** [comment]
I was thinking about a reference implementation with a scope (e.g. your iota-ed type) added as a state variable in the generated model.
Problem (as always) is the cascading calling of Validate()...

**Comment by fredbi** [comment]
Those are just free-running thoughts...

**Comment by youyuanwu** [comment]
This thread https://github.com/go-swagger/go-swagger/issues/936 and https://github.com/go-swagger/go-swagger/issues/1063 both requests the support of readOnly required by swagger 2.0 spec.

I am making an attempt to implement this feature, and I think I better share the approaches here before go into coding. 

Here are some of my observations: (using the todo-list example)
- each field has a validation function. ex. `func (m *Item) validateDescription(formats strfmt.Registry) error `
- each model struct has a validation function. ex. `func (m *Item) Validate(formats strfmt.Registry) error `

As mentioned in the thread above, the operation type needs to be accessible during the validation.
I think there are two viable approaches: 

- One way to achieve this is already mentioned above:
 `Validate(strfmt.Registry, bool disallowReadOnly) error` or more generally
` Validate(strfmt.Registry, ctx validate.Context) error`.

In ctx we can pass the operation type Read or Write, or bool IsRequest. This requires to add/pass the ctx to all validation functions, for struct and fields. We can set the ctx in  `func (o *AddOneParams) BindRequest(r *http.Request, route *middleware.MatchedRoute) error`  to be of Wirte/Request.
And we need to add to "github.com/go-openapi/validate" pkg function 
`func ReadOnly(ctx someType, path, in string, data interface{}) error`
We may need to change `Required()` function too. 

- Another way is to add meta data in each model generated. (proto-buf is doing this)
ex. 

```
// Item item
//
// swagger:model item
type Item struct {
       // store extra info for validation. And hide it from json marshal
       MetaData  someobj `json:"-"`

	// completed
	Completed bool `json:"completed,omitempty"`

	// description
	// Required: true
	// Min Length: 1
	// ReadOnly
	Description *string `json:"description"`

	// id
	// Read Only: true
	ID int64 `json:"id,omitempty"`
}
```

We can again set the metaData in `BindRequest` function.
We then need to change go-openapi/validate pkg to make use of the metaData. Advantage of this is that the existing function signatures do not need to be changed.

It is unclear to me how the approach mentioned by @fredbi above works, especially how to get the operation type when validating each field.

One question:
From swagger 2.0 spec, readOnly is defined as ` Declares the property as "read only". This means that it MAY be sent as part of a response but MUST NOT be sent as part of the request. Properties marked as readOnly being true SHOULD NOT be in the required list of the defined schema. Default value is false.`
As far As I know go-swagger validates model format of client request on the server side only (correct me if I am wrong). So is it sufficient to check request coming into the server and always reject requests with such readOnly field? Of course this means that if the struct has the ReadOnly field set, Validate will give error. But we do not validate the returned struct from server, so is this acceptable?

Let me know if there are better approaches and if I should attempt either of the approach mentioned above.

**Comment by casualjim** [comment]
Yeah I think this makes sense in the way you describe it. It would be good if it could just use context.Context.
The context paves the way for more things in the future too perhaps.

**Comment by youyuanwu** [comment]
I am thinking about context.Context too. An obvious future extension is writeOnly validation.

**Comment by fredbi** [comment]
I disagree with the change of interface.

Plus, so far, models have been standalone objects that know if they are valid or not, even out of the context of a request or some unmarshalling.

My 2 cents on this is that we could do:
- either by adding a new _extended_ validation method, e.g. `ValidateContext(context.Context, strfmt.Registry)` and
by checking in the generated code (when a read-only is at stake) that the type implements this method before calling it (with type assertion on the interface for this func)
- either by defining an unexported `operationType` in structs with at least one read-only property and using `WithOperationType` as examplified above to set the "read-only" vs "write-only" mode in a given struct before validation.

**Comment by youyuanwu** [comment]
Using unexported operationType in structs may be more trouble some.
Suppose
```go
type A struct{
    operationType string
    Content B
}

type B struct{
    operationType string
    ReadOnlyField string
}
```
When you want to validate type A, in bind request, you can set body.WithOperationType("request"), but this method needs to in turn set body.Content.WithOperationType("request"). This recursive syntax is more difficult to generate than code using context.Context.

**Comment by youyuanwu** [comment]
```go
ValidateContext(context.Context, strfmt.Registry)
```
and 
```go
Validate(strfmt.Registry)
```
co-exist there is a difficulty in combining them
Shown below:
```go
func (*m) ValidateContext(context.Context, strfmt.Registry){
    // context relavent validation goes here
   validate.ReadOnly(ctx...)
   // old validate  
   m.Validate()
}

func(*m) Validate(strfmt.Registry){
    // existing validate logic 
    m.FieldOne.Validate(ctx ???)
}
```
It is not clear how recursively validate fields of a struct can work.  

---------------------------------------------------------------
We can generate ValidateContext(context.Context, strfmt.Registry) and generate the recursive logic to validate the fields as follows:
```go
func (*m) ValidateContext(context.Context, strfmt.Registry){
   
    if v,ok := m.Field1(ValidateContext-Interface); ok{
        v.ValidateContext(ctx, formats)
    }
    // field2 ...
    // field3...
}
```
This seems quite inefficient.

@fredbi  We need to design how to change this, and it does't seem easy if we were to preserve the existing interface.

**Comment by youyuanwu** [comment]
`ValidateContext(context.Context, strfmt.Registry)` can potentially work as follows:

For each model we generate Validate(strfmt.Registry) function as before.
If the model (or its field's field...) has a readonly validation, we generate ValidateContext.

So all context independent validations go into Validate as before. and all context related validation go into ValidateContext. 
To Validate a model, we need to first call Validate and then type assert to check if it implements ValidateContext and then call ValidateContext.  (Or we can always generate those two functions to avoid type assert)

Basically the algorithm to generation of ValidateContext will be exactly the same as that of current generation of Validate, but only does context related validation.

I am trying this out.

**Comment by fredbi** [comment]
ValidateContext(context.Context, strfmt.Registry) addition merged a long time ago. Closing this old conversation.
